pkg/proxy/ipvs/graceful_termination.go

```diff
@@ -81,6 +81,9 @@ func (q *graceTerminateRSList) remove(rs *listItem) bool {
 }
 
 func (q *graceTerminateRSList) flushList(handler func(rsToDelete *listItem) (bool, error)) bool {
+	q.lock.Lock()
+	defer q.lock.Unlock()
```

Comment:
you are deadlocking, q.remove() calls the lock , L96

Reply:
Thank you @aojea, I apologize for suggesting an incorrect solution, I didn't notice that `q.remove()` also tries to acquire the lock

Reply:
no need to apologize, thanks for helping, this one was tricky

Reply:
@aojea I have been thinking of approaching this issue this way
`func (q *graceTerminateRSList) flushList(handler func(rsToDelete *listItem) (bool, error)) bool {
    q.lock.Lock()
	success := true
	for name, rs := range q.list {
		deleted, err := handler(rs)
		if err != nil {
			klog.Errorf("Try delete rs %q err: %v", name, err)
			success = false
		}
		if deleted {
			klog.Infof("lw: remote out of the list: %s", name)
                         q.lock.Unlock()
			q.remove(rs)
                         q.lock.Lock()
		}
	}
    // a check to unlock if lock is still acquired 
       if q.lock.myKey != "\x00" {
              q.lock.Unlock()
       }
	return success
}`

But I am thinking if I unlock just before return then will the function get a chance to return for sure. Because chances are that as soon as we Unlock, there might be another function that might acquire lock quickly. What do you suggest?

Reply:
Would it be possible to change the mutex to a [`RWMutex`](https://golang.org/pkg/sync/#RWMutex)?

I tried this out and it seems to work: https://play.golang.org/p/e7xaO9FYBXw
Also ran it with `-race`

Reply:
> But I am thinking if I unlock just before return then will the function get a chance to return for sure. Because chances are that as soon as we Unlock, there might be another function that might acquire lock quickly. What do you suggest?

yeah, that can be racy, I think that better holding the lock during all the operation

> RWMutex 

I was advised to not use them until you have a lot of readers, it seems there are some performance problems with the implementation

I see 2 options:
1. or just add a comment and replace q.remove(s) by its logic without the lock
```
        // same as q.remove(s) but without holding the lock
	uniqueRS := rs.String()
	if _, ok := q.list[uniqueRS]; ok {
		delete(q.list, uniqueRS)
		return true
	}
	return false
```
2. modify remove to not hold the lock and put a comment saying that the method requires the lock to be acquired, and modify the rest of code to hold the lock before calling remove

```
// remove remove an element from the rsList
// it expects the caller holds the lock
func (q *graceTerminateRSList) remove(rs *listItem) bool {
```

Reply:
> I was advised to not use them until you have a lot of readers, it seems there are some performance problems with the implementation

I see, that's fair, thanks 👍🏻 

> or just add a comment and replace q.remove(s) by its logic without the lock

I like this one a little bit better mainly because if I were to use the `remove()` method, I'd expect it work without additional logic on the caller end. 

However, if we do end up going with:
> modify remove to not hold the lock and put a comment saying that the method requires the lock to be acquired, and modify the rest of code to hold the lock before calling remove

Maybe in addition to the comment, the function name could be changed to `unsafeRemove` to make it as clear as possible, and the one other place where this would need to be changed along with holding the lock before calling this would be here:
https://github.com/kubernetes/kubernetes/blob/7e00f5d401df08ea4c0c8982594fedade3a1118b/pkg/proxy/ipvs/graceful_termination.go#L200

lmk what y'all think :)

Reply:
IMO "flush" is another form of "delete" so go for alternative 1. Copy the logic, but _not_ the returns!
```go
		if deleted {
			klog.Infof("lw: remote out of the list: %s", name)
			uniqueRS := rs.String()
			if _, ok := q.list[uniqueRS]; ok {
				delete(q.list, uniqueRS)
			}
		}
```
And I argee, do not use RWMutex or any "recursive" mutex.